sycl/test/devicelib/math_utils.hpp

```diff
@@ -1,25 +1,37 @@
 #ifndef MATH_UTILS
 #include <complex>
 #include <limits>
-using namespace std;
-// T must be float-point type
+
+// Since it is not proper to compare float point using operator ==, this
+// function measures whether the result of cmath function from kernel is
+// close to the reference and machine epsilon is used as threshold in this
+// function. T must be float-point type.
 template <typename T>
 bool is_about_FP(T x, T y) {
-  if (x == y)
-    return true;
+  bool ret;
+
+  // At least one input is nan
+  if (std::isnan(x) || std::isnan(y))
+    return std::isnan(x) && std::isnan(y);
+
+  // At least one input is inf
+  if (std::isinf(x) || std::isinf(y))
+    ret = (x == y);
+
+  // two finite
   else {
+    T threshold = std::numeric_limits<T>::epsilon() * 100;
     if (x != 0 && y != 0) {
-      T max_v = fmax(abs(x), abs(y));
-      return (abs(x - y) / max_v) <
-              numeric_limits<T>::epsilon() * 100;
-    }
-    else {
+      T max_v = std::fmax(std::abs(x), std::abs(y));
+      return std::abs(x - y) < threshold * max_v;
+    } else {
       if (x != 0)
-        return abs(x) < numeric_limits<T>::epsilon() * 100;
+        ret = std::abs(x) < threshold;
       else
-        return abs(y) < numeric_limits<T>::epsilon() * 100;
+        ret = std::abs(y) < threshold;
     }
   }
+  return ret;
```

Comment:
In this kind of code it is probably clearer to remove `ret` and just have as many early return as required.

Reply:
Thank for your comment, early return is applied to new commit and approx_equal_fp is used as the name of float point comparison.